api/v1/api_test.go

```diff
@@ -0,0 +1,273 @@
+			uniqueMetrics := metrics{
+				Type:      metricFamily.GetType().String(),
+				Help:      metricFamily.GetHelp(),
+				Timestamp: metricValues.Timestamp,
+				Metrics:   makeEncodableMetrics(metricFamily.GetMetric(), metricFamily.GetType()),
+			}
+			metricResponse[name] = uniqueMetrics
+		}
+		res = append(res, metricResponse)
+	}
+	return res
+}
+
+func compareMaps(mainMap map[string]interface{}, testMap map[string]string) bool {
+	if len(mainMap) != len(testMap) {
+		return false
+	}
+
+	for key, _ := range mainMap {
+		if mainMap[key].(string) != testMap[key] {
+			return false
+		}
+	}
+	return true
+}
+
+func compareArray(a, b []byte) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for i, v := range a {
+		if v != b[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func TestStatusAPI(t *testing.T) {
+	dms := storage.NewDiskMetricStore("", 100*time.Millisecond, nil, logger)
+	TestApi := New(logger, dms, testFlags, testBuildInfo)
+
+	req, err := http.NewRequest("GET", "http://example.org/", &bytes.Buffer{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	w := httptest.NewRecorder()
+
+	testResponse := response{}
+	TestApi.status(w, req)
+	json.Unmarshal(w.Body.Bytes(), &testResponse)
+	jsonData := testResponse.Data.(map[string]interface{})
+	responseFlagData := jsonData["flags"].(map[string]interface{})
+	responseBuildInfo := jsonData["build_information"].(map[string]interface{})
+
+	if expected, got := http.StatusOK, w.Code; expected != got {
+		t.Errorf("Wanted status code %v, got %v.", expected, got)
+	}
+
+	if !compareMaps(responseFlagData, testFlags) {
+		t.Errorf("Wanted following flags %v, got %v.", testFlags, responseFlagData)
+	}
+
+	if !compareMaps(responseBuildInfo, testBuildInfo) {
+		t.Errorf("Wanted following build info %v, got %v.", testBuildInfo, responseBuildInfo)
+	}
+}
+
+func TestMetricsAPI(t *testing.T) {
+	dms := storage.NewDiskMetricStore("", 100*time.Millisecond, nil, logger)
+	TestApi := New(logger, dms, testFlags, testBuildInfo)
+
+	req, err := http.NewRequest("GET", "http://example.org/", &bytes.Buffer{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	w := httptest.NewRecorder()
+
+	TestApi.metrics(w, req)
+
+	if expected, got := http.StatusOK, w.Code; expected != got {
+		t.Errorf("Wanted status code %v, got %v.", expected, got)
+	}
+
+	requiredResponse, _ := json.Marshal(&response{
+		Status: statusSuccess,
+		Data:   []interface{}{},
+	})
+
+	if !compareArray(requiredResponse, w.Body.Bytes()) {
+		t.Errorf("Wanted response %v, got %v.", string(requiredResponse), w.Body.String())
+	}
+
+	testTime := time.Now()
+
+	errCh := make(chan error, 1)
+
+	dms.SubmitWriteRequest(storage.WriteRequest{
+		Labels:         grouping1,
+		Timestamp:      testTime,
+		MetricFamilies: metricFamiliesMap(mf1, mf2),
+		Done:           errCh,
+	})
+
+	for err := range errCh {
+		t.Fatal("Unexpected error:", err)
+	}
+
+	w = httptest.NewRecorder()
+
+	TestApi.metrics(w, req)
+
+	requiredResponse, _ = json.Marshal(&response{
```

Comment:
As for the other, simpler example above, I'd hardcode the expected JSON string here and then compare to `w.Body.String()`. (There maybe parts in the JSON response that's not static, like timestamps. You could either leave them out (e.g. via a regexp match rather than a direct string comparison) or fill them in, making the hardcoded string a bit less hardcoded.)

As described below, the problem is that some code paths are now the same in the test and in the code to be tested, so that they will come up with the same result even if the outcome is not what we want.